ironfish/src/network/transactionFetcher.ts

```diff
@@ -0,0 +1,181 @@
+ * hashes to full transactions from the network. It operates as a state machine. Each transaction
+ * hash has it's own state which changes when peers are queried, requests timeout or a
+ * full transaction is received from the network
+ */
+export class TxFetcher {
+  // State of the current requests for each transaction
+  private readonly pending = new BufferMap<TxState>()
+
+  // Set of peers that also may be able to fetch the transaction
+  private readonly sources = new BufferMap<Set<Peer>>()
+
+  private readonly memPool: MemPool
+
+  constructor(memPool: MemPool) {
+    this.memPool = memPool
+  }
+
+  /* Called when a new transaction hash is received from the newtork
+   * This schedules requests for the hash to be sent out and if
+   * requests are already in progress, it adds the peer as a backup source */
+  addSource(hash: TransactionHash, peer: Peer): void {
+    if (this.isResolved(hash)) {
+      this.removeState(hash)
+      return
+    }
+
+    const currentState = this.pending.get(hash)
+
+    if (currentState && currentState.peer === peer) {
+      return
+    }
+
+    const sources = this.sources.get(hash) || new Set<Peer>()
+    sources.add(peer)
+    this.sources.set(hash, sources)
+
+    if (!currentState) {
+      this.fetchFromNextSource(hash)
+    }
+  }
+
+  /* Called when a transaction is received and confirmed from the network
+   * either in a block or in a gossiped transaction request */
+  resolve(transaction: Transaction): void {
+    this.removeState(transaction.hash())
+  }
+
+  private fetchFromNextSource(hash: TransactionHash): void {
+    if (this.isResolved(hash)) {
+      this.removeState(hash)
+      return
+    }
+
+    // Clear the previous source
+    const currentState = this.pending.get(hash)
+    currentState && this.cleanupCallbacks(currentState)
+
+    // Get the next source
+    const peer = this.popSource(hash)
+    if (!peer) {
+      this.pending.delete(hash)
+      return
+    }
+
+    const timeout = setTimeout(() => this.sendRequest(hash, peer), 500)
+
+    this.pending.set(hash, {
+      peer,
+      action: 'REQUEST_SCHEDULED',
+      timeout,
+    })
+  }
+
+  private sendRequest(hash: TransactionHash, peer: Peer): void {
+    if (this.isResolved(hash)) {
+      this.removeState(hash)
+      return
+    }
+
+    const message = new PooledTransactionsRequest([hash])
+
+    const sent = peer.send(message)
+
+    if (!sent) {
+      this.fetchFromNextSource(hash)
+      return
+    }
+
+    const onPeerDisconnect = ({ peer, state }: { peer: Peer; state: PeerState }) => {
+      if (state.type === 'DISCONNECTED') {
+        peer.onStateChanged.off(onPeerDisconnect)
+        this.fetchFromNextSource(hash)
+      }
+    }
+    peer.onStateChanged.on(onPeerDisconnect)
+
+    const clearDisconnectHandler = () => {
+      peer.onStateChanged.off(onPeerDisconnect)
+    }
+
+    const timeout = setTimeout(() => {
+      this.fetchFromNextSource(hash)
+    }, 5000)
```

Comment:
Just noting that a lot of logic here is similar to the RPC messaging system -- I think it's okay as-is, so I'm not sure it's worth refactoring to reuse that code, but it's pretty identical but with a different timeout

Reply:
I know, I was trying to figure out a way to combine them. Especially since we try to resolve all RPC Responses in the PeerNetwork and this one would kind of be bypassing that. The main problem I was running into was that if a Transaction comes in while the request is out then we want to clear the handlers (timeout and peer onStateChanged). If we used the RPC messaging logic those would kind of be hidden from us. I'll try to think about it a little more to see if I can come up with a good solution